internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -267,67 +265,207 @@ func writeKubeadmConfig(data *userdata.UserData) (err error) {
+	var err error
+
+	trustds := []proto.TrustdClient{}
+
+	creds, err = basic.NewCredentials(data.Services.Trustd)
+	if err != nil {
+		return trustds, err
+	}
+
+	// Create a trustd client for each endpoint to set up
+	// a fan out approach to gathering the files
+	var conn *grpc.ClientConn
+	for _, endpoint := range data.Services.Trustd.Endpoints {
+		conn, err = basic.NewConnection(endpoint, constants.TrustdPort, creds)
+		if err != nil {
+			return trustds, err
 		}
+		trustds = append(trustds, proto.NewTrustdClient(conn))
 	}
-	return nil
+	return trustds, nil
 }
 
-func writeFiles(client proto.TrustdClient, files []string) (err error) {
-	errChan := make(chan error)
-	doneChan := make(chan bool)
-	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Minute)
-	defer cancelFunc()
-
-	go func() {
-		<-ctx.Done()
-		errChan <- ctx.Err()
-	}()
-
-	go func() {
-		var err error
-		for _, f := range files {
-		L:
-			req := &proto.ReadFileRequest{
-				Path: f,
+func download(client proto.TrustdClient, file *proto.ReadFileRequest, content chan<- []byte, donezo <-chan struct{}) {
+	var resp *proto.ReadFileResponse
+	var err error
+	attempt := 0
+	maxWait := float64(64)
+
+	// Wait indefinitely with a backoff
+	for {
+		attempt++
+
+		select {
+		case <-donezo:
+			return
+		default:
+		}
+
+		if resp, err = client.ReadFile(context.Background(), file); err == nil {
+			select {
+			case <-donezo:
+				return
+			case content <- resp.Data:
+			case <-time.After(2 * time.Second):
+				// retry
+				// help with expiration of unnecessary requests
 			}
-			var resp *proto.ReadFileResponse
-			if resp, err = client.ReadFile(context.Background(), req); err != nil {
-				log.Printf("failed to read file %s: %v", f, err)
-				time.Sleep(1 * time.Second)
-				goto L
+		}
+
+		if err != nil {
+			log.Printf("failed to read file %s: %+v", file, err)
+		}
+
+		// backoff
+		snooze := math.Pow(2, float64(attempt))
+		if snooze > maxWait {
+			snooze = maxWait
+		}
+		time.Sleep(time.Duration(snooze) * time.Second)
+	}
+}
+
+// writeTrustdFiles handles reading the replies from trustd and writing them
+// out to a file on disk
+func writeTrustdFiles(requestedFile string, content chan []byte, donezo chan struct{}) {
+	// Wait for our responses
+	for {
+		select {
+		case data := <-content:
+			// If the file already exists, no need to write it again
+			// TODO: perhaps look at some sort of checksum verification?
+			if _, err := os.Stat(requestedFile); err == nil {
+				continue
 			}
-			if err = ioutil.WriteFile(f, resp.Data, 0400); err != nil {
-				log.Printf("failed to read file %s: %v", f, err)
-				time.Sleep(1 * time.Second)
-				goto L
+
+			// Unsure on the perms for this
+			// NOTE for review, we need to agree on perms
+			// ref writeKubeadmPKIFiles()
+			if err := os.MkdirAll(filepath.Dir(requestedFile), 0755); err != nil {
+				log.Printf("failed to create directory for %s: %v", requestedFile, err)
+				continue
```

Comment:
I'm not sure I understand the flow here, probably missing something.

Looks like downloader will send data (the same each time?) to `content` channel and here we retry if we can't create a directory? how this error can be cleared with the next download?

Reply:
I ran into an issue where it was unable to write out the fetched file because the destination path did not exist so we need to create it. Honestly I'm not sure what to do in this situation if there's an error creating the directory. Perhaps it should be a hard fail?